src/deepsparse/loggers/function_logger.py

```diff
@@ -0,0 +1,146 @@
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Implementation of the Function Logger
+"""
+from collections import defaultdict
+from pathlib import Path
+from typing import Any, Dict, List, Union
+
+import yaml
+
+from deepsparse.loggers import BaseLogger, MetricCategories
+from deepsparse.loggers.metric_functions import apply_function
+
+
+__all__ = ["FunctionLogger"]
+
+
+class FunctionLogger(BaseLogger):
+    """
+    DeepSparse logger that applies functions to raw log values
+    according to the specified config file
+
+    :param logger: A DeepSparse Logger object
+    :param config: A configuration dictionary that specifies the mapping between
+        the target name and the functions to be applied to raw log values
+
+        `config` can be specified as a dictionary type:
+        e.g.
+
+        {"pipeline_inputs":
+            [{"function": "some_function_1",
+              "frequency": 3},
+
+            [{"function": "some_function_2",
+              "frequency": 5}],
+
+         "pipeline_outputs":
+            [{"function": "some_function_1",
+            "frequency": 4}]
+        }
+
+        or a a yaml file:
+        e.g.
+
+        ```yaml
+        pipeline_inputs:
+            - function: some_function_1
+              frequency: 3
+            - function: some_function_2
+              frequency: 5
+        pipeline_outputs:
+            - function: some_function_1
+              frequency: 4
+        ```
+    """
+
+    def __init__(
+        self,
+        logger: BaseLogger,
+        config: Union[str, Dict[str, List[Dict[str, Any]]]],
+    ):
+
+        self.logger = logger
+        self.config = (
+            config
+            if isinstance(config, dict)
+            else yaml.safe_load(Path(config).read_text())
+        )
+        self.function_call_counter = self._create_function_call_counter(self.config)
+
+    def log(self, identifier: str, value: Any, category: MetricCategories):
+        """
+        :param identifier: The name of the thing that is being logged.
+        :param value: The data structure that the logger is logging
+        :param category: The metric category that the log belongs to
+        """
+        if category == MetricCategories.DATA:
+            # logging data information
+            pipeline_name, *target = identifier.split(".")
+            self._log_data(
+                pipeline_name=pipeline_name,
+                target=".".join(target),
+                value=value,
+                category=category,
+            )
+        else:
+            # logging system information
```

Comment:
we still want to support functions for non data functions, at the server level we'll just add in defaults to enable Identity calls for them